Query refactoring: BoolQueryBuilder and Parser

cbuescher opened the pull request:
Refactors BoolQueryBuilder and Parser. Splits the parse() method into a parsing and a query building part, adding NamedWriteable implementation for serialization and hashCode(), equals() for testing.

This change also adds tests using fixed set of leaf queries (Terms, Ids, MatchAll) as nested Queries in test query setup. Also QueryParseContext is adapted to return QueryBuilder instances for inner filter parses instead of previously returning Query instances, keeping old methods in place but deprecating them.

Relates to #10217

PR goes agains query-refacoring feature branch.

cbuescher commented on src/main/java/org/elasticsearch/index/query/BaseQueryBuilder.java:
```diff
@@ -22,17 +22,30 @@
  */
 public abstract class BaseQueryBuilder implements QueryBuilder {
 
+    private final static Map<String, Class<?>> queryBuilderRegistry = new HashMap<>();
```
This map should eventually be replaced by registry for the Builders (or whatever we call them in the end) that we have today with the IndicesQueriesRegistry, only that this looks like a registry for the Parsers. I've introduced this static map here for lack of a better solution, but if we eventually merge parsers and builders and can access them this would be good to do here. Any other suggestions?

MaineC replied:
If we merge parsers and builders at some point anyway why not re-use the registry for query parsers we already have in place and add a "get builder" method to the existing query parsers?

Something along the lines of:
    public abstract class BaseQueryBuilder implements QueryBuilder {
        IndexQueryParserService service;
    
        @Inject
        public BaseQueryBuilder(IndexQueryParserService service) {
            this.service = service;
        }

        protected QueryBuilder readQueryBuilderFrom(StreamInput in) throws IOException {
            QueryBuilder newBuilder;
            String name = in.readString();

            // the following two lines would become one once parsers and builders are merged
            QueryParser parser = this.service.queryParser(name);
            QueryBuilder builder = parser.createBuilder();
            builder.readFrom(in);
        }

cbuescher replied:
Like this in general, will start looking into this. However the service / lookup field needs to somehow be static since it's read-only and we need it for every query deserialization, I think.

javanna replied:
The only thing about this approach is the dependency that we would create between parser and builder... imagine the request comes from the java API, it's in binary format already, and we have to go to the parser to ask for an instance of the builder, but no parsing is needed. That feels slightly weird, maybe it's not a bit problem though.

javanna replied:
about my last comment, along the same lines, that is why I don't love the parserName method in the builders, and I hope it goes away soon (I thought at first it would be temporary, but maybe it'll have to stay, we'll see). Ideally, builders don't know about parsers, but parsers do know about their respective builders. What do you folks think about this?

cbuescher replied:
I think this makes sense, since the Builders can exist without the other (as you said, going through the java API only). The link is the NAME field at the moment. What if in a first step we reverse that direction like you mentioned:
* move all unique NAME constants from the parsers to the builders
* getParserName() in the Builder classes goes away since its known then.
* I think the names() method in the parsers has to stay for now since its used to register them
* as long as we have parser-registry we could add a createBuilder() method to each parser that returns an empty builder instance. Or even stream into that.

I would do that in a separate PR though.

MaineC replied:
@javanna Just for clarification - my proposal only makes sense if ultimately we intend to merge parsers and builders. Otherwise to me it feels like a hack.

@cbuescher +1

javanna replied:
@MaineC yea I see, to be honest I don't know yet if we'll merge those at the end, I kinda like the separation between the two, but we will see. I like Christoph's idea around moving the name to the builder, then the parser can re-use it too rather than the other way around.

>I think the names() method in the parsers has to stay for now since its used to register them

sure but it can always return something taken from the corresponding builder

> as long as we have parser-registry we could add a createBuilder() method to each parser that returns an empty builder instance. Or even stream into that.

ah I see, so you would avoid registering two things, cause the parser can provide both, and they both need to be registered with the same name. This sounds good to me. Would make it easier for plugins to migrate and keep the registration of custom queries to one single call. Just to clarify, this wouldn't be a create* method, cause it wouldn't create an instance all the time, but simply return always the same empty builder, right?

javanna commented on src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -145,24 +242,27 @@ public BoolQueryBuilder queryName(String queryName) {
-        if (disableCoord != null) {
-            builder.field("disable_coord", disableCoord);
-        }
+        builder.field("disable_coord", disableCoord);
```
not sure about always printing this out, why?

javanna commented on src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -145,24 +242,27 @@ public BoolQueryBuilder queryName(String queryName) {
-        if (adjustPureNegative != null) {
-            builder.field("adjust_pure_negative", adjustPureNegative);
-        }
+        builder.field("adjust_pure_negative", adjustPureNegative);
```
not sure about always printing this out too, why?

javanna commented on src/main/java/org/elasticsearch/index/query/BaseQueryBuilder.java:
```diff
@@ -85,11 +99,61 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
+    return newBuilder;
+}
+
+    protected static void writeQueryBuilderTo(StreamOutput out, QueryBuilder queryBuilder) throws IOException {
```
I think this should rather be in StreamOutput? what do you think?

javanna commented on src/main/java/org/elasticsearch/index/query/BaseQueryBuilder.java:
```diff
@@ -85,11 +99,61 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
+        return list;
+    }
+
+    protected static void writeQueryBuilderListTo(StreamOutput out, List<QueryBuilder> list) throws IOException {
```
I think this should rather be in StreamOutput? what do you think?

javanna commented on src/main/java/org/elasticsearch/index/query/BaseQueryBuilder.java:
```diff
@@ -85,11 +99,61 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
+        queryBuilder.writeTo(out);
+    }
+
+    protected static List<QueryBuilder> readQueryBuilderListFrom(StreamInput in) throws IOException {
```
move to StreamInput?

javanna commented on src/main/java/org/elasticsearch/index/query/BaseQueryBuilder.java:
```diff
@@ -85,11 +99,61 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
+        Class<?> newBuilderClass = queryBuilderRegistry.get(name);
+        if (newBuilderClass != null) {
+            try {
+                newBuilder = (QueryBuilder) newBuilderClass.newInstance();
```
reflection... I'd try to avoid it if possible. 

Can we move from `Streamable` to `StreamableReader` for all query builders? `readFrom` would then create a new instance internally. In the registry we would just need empty objects, we don't need to create them all the time with reflection, just have a Map<String,QueryBuilder>, the lookup already returns an empty instance that we can use to call readFrom against, which will return the actual de-serialized object. Makes sense?

javanna replied:
Wonder if this can be moved to StreamInput somehow, but not sure where the registry would go then....

javanna commented:
I left a few comments. I think I would split this into two PRs, one for adding Streamable support to all query builders, and streamline serialization support, second one around actually refactoring the bool query.

As I said in my comments I think we should move all query builders to `StreamableReader` and change some of the existing query builders to have final fields when it make sense, lose the default constructors when possible etc.

cbuescher commented:
Did a rebase and squash after merge of NamedWritable interface on feature branch. I think this is good for another round of reviews now.

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -93,6 +149,25 @@ public BoolQueryBuilder should(QueryBuilder queryBuilder) {
+    }
+
+    /**
+     * Gets the queries that <b>should</b> appear in the matching documents.
```
can't quite read it. how about "Returns the clauses that should be matched by the returned documents" ? in general should appear is probably not the best way to say it (I'm not a native english speaker though!)

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -169,18 +289,14 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
-        if (adjustPureNegative != null) {
-            builder.field("adjust_pure_negative", adjustPureNegative);
-        }
+        builder.field("adjust_pure_negative", adjustPureNegative);
```
I am starting to see that the default boost doesn't get printed out but other default fields do. Makes sense to me but maybe we want to be consistent? I think we should have this a separate discussion, make some decision and do the same everywhere (I have the feeling we are not yet settled yet on one way or another)

cbuescher replied:
Agreed on having separat discussion for this. Should I open an issue?

javanna replied:
no issue, lets talk about it tomorrow and reach some agreement? :)

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -207,4 +323,96 @@ private void doXArrayContent(String field, List<QueryBuilder> clauses, XContentB
+        return null;
+    }
+
+    private void addBooleanClauses(QueryParseContext parseContext, BooleanQuery booleanQuery, List<QueryBuilder> clauses, Occur occurs)
```
could this be static?

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -207,4 +323,96 @@ private void doXArrayContent(String field, List<QueryBuilder> clauses, XContentB
+        QueryBuilder.writeQueryBuilderListTo(out, this.mustClauses);
+        QueryBuilder.writeQueryBuilderListTo(out, this.mustNotClauses);
+        QueryBuilder.writeQueryBuilderListTo(out, this.shouldClauses);
+        QueryBuilder.writeQueryBuilderListTo(out, this.filterClauses);
```
you don't need to specify QueryBuilder. do you?

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -207,4 +323,96 @@ private void doXArrayContent(String field, List<QueryBuilder> clauses, XContentB
+        boolQueryBuilder.adjustPureNegative = in.readBoolean();
+        boolQueryBuilder.disableCoord = in.readBoolean();
+        boolQueryBuilder.minimumShouldMatch = in.readOptionalString();
+        return boolQueryBuilder;
```
here too, similar to my doubts around printing out default vaues. In general, shall we have some discussion on whehter we want to always serialize everything. Sounds good to me, I just wonder if we should look into serializing less fields given that most of the times everything will be default and we could save bandwidth? Or maybe this is anyways not a problem at all?

javanna replied:
I don't mean to block this PR specifically, these are just things that are popping up the more I review similar PRs

cbuescher replied:
In that case (e.g. for boost), why do we currently use primitives instead of Objects (e.g. Float). In that case it would be clear if a value was provided by the user or we should use the default. Just curious what the reasons for using primitives in many places like this are.

javanna replied:
I guess we just prefer primitive types over objects :)

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryBuilder.java:
```diff
@@ -120,4 +122,23 @@ public QB readFrom(StreamInput in) throws IOException {
+        for (QueryBuilder queryBuilder : list) {
+            out.writeNamedWriteable(queryBuilder);
+        }
+    }
```
shall we add readNamedWriteableList and writeNamedWriteableList to StreamInput and StreamOutput? seems like it could be reused in general and doesn't necessarily belong here.

cbuescher replied:
Since it only deals with lists of QueryBuilders I found it more natural to keep it here in the base class for all QueryBuilders rather than blow up StreamInput. I think I saw similar approaches e.g. in the Rounding serialization. But if it is useful elsewhere we can always make it public?

javanna replied:
the point is that it is not only about QueryBuilders. it is just about NamedWriteables, could be used somewhere else too indeed

cbuescher replied:
Great, got that. Will do.

javanna commented on core/src/test/java/org/elasticsearch/index/query/BoolQueryBuilderTest.java:
```diff
@@ -0,0 +1,136 @@
+        for (int i = 0; i < randomIntBetween(0, 3); i++) {
+            query.should(createRandomQuery());
+        }
+        for (int i = 0; i < randomIntBetween(0, 3); i++) {
```
all these random values need to be saved out of the loop...you know what happens otherwise? :)

cbuescher replied:
Arg, gotcha. Guess I was lucky so far.

javanna commented on core/src/test/java/org/elasticsearch/index/query/BoolQueryBuilderTest.java:
```diff
@@ -0,0 +1,136 @@
+        return query;
+    }
+
+    private static QueryBuilder createRandomQuery() {
```
shall we move this method to a new QueryTestUtils class or something? I see it being used in other tests in the future and I am not sure it belongs to the test base class

cbuescher replied:
I tried moving this to a separate class but since method uses randomness, its better situated inside a test (or make that new Util-class a test, which feels strange to me). So I moved it to the BaseQueryTestCase where it can be re-used by other Query tests.

javanna replied:
the weird thing about this solution is that the base class knows about its subclasses, that feels horribly wrong. You can always have a util class that accepts Random as an argument or something no?

javanna commented on core/src/test/java/org/elasticsearch/index/query/BoolQueryBuilderTest.java:
```diff
@@ -0,0 +1,136 @@
+                luceneQuery = new TermQueryBuilderTest().createExpectedQuery((TermQueryBuilder) query, parseContext);
+            }
+            if (query instanceof IdsQueryBuilder) {
+                luceneQuery = new IdsQueryBuilderTest().createExpectedQuery((IdsQueryBuilder) query, parseContext);
```
along the same lines as my comment above, not sure how much we want to complicate this already, but maybe we want to introduce a new RandomQuery construct that exposes createTestQueryBuilder and createExpectedQuery ? then it could be used in both this test and the test that we depend on here? Also if we did this for each query we could expand testing? How does it sound?

cbuescher replied:
Rather than setting up a whole new system here I decided to rely on toQuery() for all nested queries (after all we test the main QueryBuilder here) in creating the expected BoolQuery and not use createExpectedQuery here. Let me know what you think, we can alway extend the scope here.

javanna replied:
yea I think it makes sense, indeed we shoudn't complicate this too much, leaf queries get their own tests anyways.

javanna commented:
looks good, left some comments

cbuescher commented:
Adressed last round of comments, putting reading/writing of NamedWritable lists in the Streaming classes amongst other things. Let me know if you think this is good to go.

javanna commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -509,6 +509,22 @@ public Object readGenericValue() throws IOException {
+        int size = readInt();
+        for (int i = 0; i < size; i++) {
+            C obj = readNamedWriteable();
+            if (obj != null) {
```
can it ever be null? I think there is something wrong if it is null, cause it should always throw exception upfront in case of issues?

javanna commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -509,6 +509,22 @@ public Object readGenericValue() throws IOException {
+     * Reads a list of {@link NamedWriteable} from the current stream, by first reading its size and then
+     * reading the individual objects using {@link #readNamedWriteable()}.
+     */
+    public <C> List<C>  readNamedWritableList() throws IOException {
```
one too many whitespace between List<C> and the method name

javanna commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -457,4 +457,15 @@ public void writeNamedWriteable(NamedWriteable namedWriteable) throws IOExceptio
+     * Writes a list of {@link NamedWriteable} to the current stream, by first writing its size and then iterating over the objects
+     * in the list
+     */
+    public void writeNamedWritableList(List<? extends NamedWriteable> list) throws IOException {
```
thinking out loud, should we support Collection instead of List in both methods?

cbuescher replied:
In write that would be okay, but to be consistent we should then also allow Collection in read and that means we need to pass in an empty Collection to the read-method. I'm not sure if this causes more trouble than necessary, you might accidentally mix orderings etc...
I would leave it for now an generalize if the need arises?

javanna replied:
makes sense, lets leave list for now

javanna commented:
left a couple of tiny comments, LGTM otherwise

Commit:
Refactors BoolQueryBuilder and Parser. Splits the parse(QueryParseContext ctx) method into a parsing
and a query building part, adding NamedWriteable implementation for serialization and hashCode(),
equals() for testing.

This change also adds tests using fixed set of leaf queries (Terms, Ids, MatchAll) as nested Queries in test query setup.
Also QueryParseContext is adapted to return QueryBuilder instances for inner filter parses instead of
previously returning Query instances, keeping old methods in place but deprecating them.

Relates to #10217
Closes #11427
